### ISSUE TITLE
Why is domain() necessary?

### ISSUE THREAD

**Issue by zsunberg** [sisl/POMDPs.jl#47]
I can't remember if we've discussed why domain() is in the interface. Why not just define start(), next(), and done() directly for subtypes of AbstractSpace?

The only reason I can think of is that it might be convenient to just have domain() return an array in some cases. Is there something else?

**Comment by mykelk** [comment]
For continuous action spaces.

**Comment by zsunberg** [comment]
Hmm... not sure I follow. What would domain() return for a continuous space?

I think both continuous and discrete spaces could be handled without the
need for domain().

**Comment by mykelk** [comment]
Domain could return an object of a type like this:
```julia
type ContinuousRange <: Domain
    lowerbound
    upperbound
end
```
How would you do it without `domain()`?

**Comment by zsunberg** [comment]
Perhaps I am misunderstanding what the "iterator" that domain() is supposed
to return is. I am under the impression that an "iterator" is just an
object that can be used in a for loop
```julia
for a in domain(my_action_space)
    # do something with a
end
```
How could one iterate through the elements of a `ContinuousRange` object?
It has uncountably many elements (ignoring the binary representation of
course).

**Comment by mykelk** [comment]
I was thinking that a domain would be a more general concept. If you can iterate through it, then great. If not, then you'd have to dispatch based on the type that it returns. That was my original thinking, but it may be too much of an abstraction to implement for our first cut.

**Comment by zsunberg** [comment]
Yeah, I think that if we are going to have a domain be a different concept
from a space, we should nail down what that concept is. If we don't have
two different concepts, we should consider getting rid of domain() and, if
the space can be iterated through, `start()`, `next()` and `done()` should
be implemented directly for the space so that it can be used like this:
```julia
for a in my_action_space
    # do something with a
end
```

**Comment by mykelk** [comment]
That's fine with me. What do the others think?

**Comment by ebalaban** [comment]
I agree with @zsunberg. I think that for now space coupled with its iterator functions is the cleaner way to go. If we do develop domain as a distinct concept, we can revisit this.

**Comment by etotheipluspi** [comment]
I'm ok with this. I'll implement the change this weekend.

**Comment by zsunberg** [comment]
By making this change, we would be expecting new users (who may be new to julia) to understand the start(), next(), and done() interface. Is that ok? I'm just imagining new-masters-student-Zach sitting down to do a homework problem based on this in an afternoon and getting really confused. I suppose that maybe for a homework problem, we wouldn't expect them to start from scratch; we would give them some space types and have them implement some of the other functions.

**Comment by etotheipluspi** [comment]
The object doesn't have to be an iterator, but an iterable type, i.e. an array. There is no reason to require the user to implement the start(), next(), done() functions unless they know what they are doing. Beginners, for example,  could make their states(pomdp) function return an array that contains all the states of the problem

**Comment by zsunberg** [comment]
Yes... unless the solver uses `::AbstractSpace` somewhere. Is there a way
to make an array a subtype of `::AbstractSpace`

**Comment by etotheipluspi** [comment]
One option would be to use Union to create a new type like a SpaceIterable that can be used in the solvers.
```julia
SpaceIterable = Union{AbstractSpace, Array}
```
Adding a new type is not desirable, but I don't think we want to do anything with inheritance of Arrays here.

**Comment by ebalaban** [comment]
Not too crazy about the union type, but can't think of a better solution at the moment for the no-iterators case.

**Comment by zsunberg** [comment]
This is why I thought you guys had domain() in the first place, haha -
because it would be convenient to do
```julia
type MyActionSpace <: AbstractSpace
    container::Array{MyAction}
end
domain(as::MyActionSpace) = as.container
```

Now I think it's better to just require start(), next(), and done() for
spaces as I suggested above. People who want to implement their own
problems from scratch should learn how to do that. For homework problems
and the like, we can provide the space types.

**Comment by ebalaban** [comment]
I agree with @zsunberg on start(), next(), and done(). I think that provides a more consistent interface to the solvers. If one is writing their own problem definition, getting those functions implemented is not a big deal and yes, we can always provide some examples. As for the "wrapper" option, I am not terribly crazy about that one either :-)

**Comment by mykelk** [comment]
I'd rather keep domain than create a union type.

**Comment by tawheeler** [comment]
we could use a different function name for continuous spaces. Maybe extrema. It could stil be called on a domain object, but one for which you don't implement iteration.

**Comment by mykelk** [comment]
Domain is probably not the right word since domains can be continuous. Maybe iterate or something like that---that can imply a discrete sequence?

**Comment by etotheipluspi** [comment]
I would rather avoid the requirement of writing start(), next() and done(), and vote for having something like @mykelk suggested with an iterate function.

**Comment by mykelk** [comment]
Just to clarify, the iterate function I had in mind would return something that is iterable---e.g., an array.

**Comment by ebalaban** [comment]
Hm, I suppose this is fine, although having done() leaves room for implementing something clever if one does not want to allocate and iterate through an entire huge state space, for example.

**Comment by mykelk** [comment]
Wait... my suggestion wasn't to get rid of start(), next(), and done()... it was just to allow iterate to return something that is iterable such as an array or an iterator---which would include done().

**Comment by etotheipluspi** [comment]
Yes, I had the same idea as @mykelk. Which is basically what we have now, except domain() is replaced with a different function name like iterate().

iterate() returns an iterable object which may be an iterator or an array. If iterate() returns an iterator that the user wrote, it can have custom start(), next(), done() implementations. If it returns an array, thats fine too. Both get treated the same by the solvers.

**Comment by mykelk** [comment]
Unless there are strong objections, let's go with that.

**Comment by ebalaban** [comment]
Let me try to summarize the dilemma here, perhaps this will help us get on the same page. Here are the options proposed so far, as I see them:

1. Require problem writers to return an iterator type, such as MySpaceIterator <: AbstractSpace, and provide start(), next(), and done() along with it. I am in favor of this option and I am guessing that @zsunberg is in favor of this option as well.

2. Allow problem writers to return either an iterator type as in Option 1 or an iterable object of some kind, such as an array, for which start(), next(), and done() have already been predefined. This would be perfectly fine, the only problem is that an array can't be a subclass of AbstractSpace, so if some solver expects a subclass of AbstractSpace, that's not going to work.

3. Define a union type, such as SpaceIterable = Union{AbstractSpace, Array}. That would work, but is a bit unseemly, in my opinion.

4. Wrap either a custom iterator type or an iterable object, such as an array, into a "wrapper" type, e.g. MyActionSpace <: AbstractSpace. This would allow subtyping from AbstractSpace, but would require solver writers to know which field to access in an object of MyActionSpace type, e.g. my_action_space.container. I think that this is not all that elegant either.

5. Allow spaces to only be described in terms of something already iterable, such as an array. No need to implement start(), next(), and done(). This is the option that I thought @etotheipluspi  and @mykelk were proposing earlier today, but I guess they were favoring something closer to Option 2 instead.

Did I miss or misinterpret something?

**Comment by zsunberg** [comment]
For this post at least define the term "iterable" to mean that the object
has start(), next(), and done() defined for it. Here are the two options I
see above:

1. @ebalaban and I are advocating that an AbstractSpace should be iterable
if possible.

2. @mykelk and @etotheipluspi are advocating renaming domain() to iterate()
and requiring that the returned object be iterable (just as the
documentation for domain() required previously).

I slightly prefer 1, but I am ok with 2. I think that for 2 name should be
iterable(), but am not set on that.

**Comment by ebalaban** [comment]
I see, I guess I did misunderstand slightly what @mykelk and @etotheipluspi meant. So, we'd still have space types be subtypes of AbstractSpace, but use iterate(space::AbstractSpace) to return either a custom iterator type or an array, for example. That still leaves the question though of what type can we expect to get back from iterate(). Just 'Any'?

**Comment by ebalaban** [comment]
I would be ok with that second option as well, by the way, and I guess the type returned from iterate() should not really matter if we just use it in loop constructs.

**Comment by zsunberg** [comment]
exactly

**Comment by ebalaban** [comment]
Ok, sounds like we've converged then. I have a slight preference for 'iterate' as the name, but would be ok with 'iterable' as well.

**Comment by mykelk** [comment]
What is better? Iterate or iterable? Iterable sounds like a Boolean function. Iterate is used by iterators.jl, which has been merged into base in 0.4. https://github.com/JuliaLang/Iterators.jl/blob/master/README.md

**Comment by zsunberg** [comment]
true (no pun intended), iterable() indeed sounds Boolean. iterate in
Iterators.jl has a slightly different meaning than we have here.

What about iterator()? In C++ "iterator" does not refer to the container
itself, but to a pointer to an element, however, at some places in the
julia documentation, "iterator" does seem to refer to something that can be
iterated through (see e.g. the documentation for enumerate()). iterator()
also seems more consistent with our other function names, e.g. updater().

I'm ok with any of these choices.

**Comment by etotheipluspi** [comment]
I like the sound of iterable more, but it does sound like a boolean function. Also, iterate does not seem to be in base in 0.4, although many other functions from Iterators.jl are. Since iterate is not in base, I would vote against using it due to the conflicts it can cause if the user loads Iterators.jl. 

I discussed this with Tim, and we agree that having something like get_iterable would be them most clear cut name, but this goes against the Julia naming convention :(

**Comment by mykelk** [comment]
I'm okay with iterator(). I don't like get_ functions.

**Comment by etotheipluspi** [comment]
I said this above, but the reason iterator() is not a good choice is that its NOT imported into base in 0.4. So in order to use it in POMDPs.jl (avoid conflicts if user uses Iterators), we would have to explicitly export it from Iterators.jl. I don't think we want to do that.

**Comment by tawheeler** [comment]
What about `iter(::AbstractSpace)`?

**Comment by mykelk** [comment]
I'm okay with that.

**Comment by zsunberg** [comment]
@etotheipluspi, I see iterate() in Iterators.jl, but not iterator(). I am
also ok with iter()

**Comment by ebalaban** [comment]
I like iterator(). I can live with iter() as well.

**Comment by etotheipluspi** [comment]
@zsunberg you're right, I completely missed that you proposed iterator() and not iterate(). Sorry about that.

My vote is for iterator().

**Comment by tawheeler** [comment]
If it isn't used in Iterators.jl, `iterator()` would be clearer.

**Comment by ebalaban** [comment]
Since we seem to have converged on the name, I went ahead and put iterator() in (I needed it to wrap some of my code up). I replaced domain() with iterator() in the README, but still left it in the code for now. Should we just remove it altogether or redefine to do something for continuous spaces?

**Comment by mykelk** [comment]
I vote to remove for now.

**Comment by zsunberg** [comment]
Yep. Remove for now. Thanks for making the change

**Comment by etotheipluspi** [comment]
That's fine with me.

**Comment by ebalaban** [comment]
Ok, great, thanks! Implemented with a58e35a and 15ee577.
